daemon/volumes.go

```diff
@@ -1,225 +1,115 @@
 	}
 
-	return nil
+	return "", fmt.Errorf("Unable to setup mount point, neither source nor volume defined")
 }
 
-// sortedVolumeMounts returns the list of container volume mount points sorted in lexicographic order
-func (container *Container) sortedVolumeMounts() []string {
-	var mountPaths []string
-	for path := range container.Volumes {
-		mountPaths = append(mountPaths, path)
+func (m *mountPoint) Path() string {
+	if m.Volume != nil {
+		return m.Volume.Path()
 	}
 
-	sort.Strings(mountPaths)
-	return mountPaths
+	return m.Source
 }
 
-func (container *Container) VolumePaths() map[string]struct{} {
-	var paths = make(map[string]struct{})
-	for _, path := range container.Volumes {
-		paths[path] = struct{}{}
+func parseBindMount(spec string, config *runconfig.Config) (*mountPoint, error) {
+	bind := &mountPoint{
+		RW: true,
 	}
-	return paths
-}
-
-func (container *Container) registerVolumes() {
-	for path := range container.VolumePaths() {
-		if v := container.daemon.volumes.Get(path); v != nil {
-			v.AddContainer(container.ID)
-			continue
-		}
+	arr := strings.Split(spec, ":")
 
-		// if container was created with an old daemon, this volume may not be registered so we need to make sure it gets registered
-		writable := true
-		if rw, exists := container.VolumesRW[path]; exists {
-			writable = rw
-		}
-		v, err := container.daemon.volumes.FindOrCreateVolume(path, writable)
-		if err != nil {
-			logrus.Debugf("error registering volume %s: %v", path, err)
-			continue
+	switch len(arr) {
+	case 2:
+		bind.Destination = arr[1]
+	case 3:
+		bind.Destination = arr[1]
+		if !validMountMode(arr[2]) {
+			return nil, fmt.Errorf("invalid mode for volumes-from: %s", arr[2])
 		}
-		v.AddContainer(container.ID)
+		bind.RW = arr[2] == "rw"
+	default:
+		return nil, fmt.Errorf("Invalid volume specification: %s", spec)
 	}
-}
 
-func (container *Container) derefVolumes() {
-	for path := range container.VolumePaths() {
-		vol := container.daemon.volumes.Get(path)
-		if vol == nil {
-			logrus.Debugf("Volume %s was not found and could not be dereferenced", path)
-			continue
+	if !filepath.IsAbs(arr[0]) {
+		bind.Driver, bind.Name = parseNamedVolumeInfo(arr[0], config)
+		if bind.Driver == volume.DefaultDriverName {
+			return nil, localMountErr
 		}
-		vol.RemoveContainer(container.ID)
+	} else {
+		bind.Source = filepath.Clean(arr[0])
 	}
-}
 
-func parseBindMountSpec(spec string) (*volumeMount, error) {
-	arr := strings.Split(spec, ":")
+	bind.Destination = filepath.Clean(bind.Destination)
+	return bind, nil
+}
 
-	mnt := &volumeMount{}
-	switch len(arr) {
+func parseNamedVolumeInfo(info string, config *runconfig.Config) (driver string, name string) {
+	p := strings.SplitN(info, "/", 2)
+	switch len(p) {
 	case 2:
-		mnt.hostPath = arr[0]
-		mnt.containerPath = arr[1]
-		mnt.writable = true
-	case 3:
-		mnt.hostPath = arr[0]
-		mnt.containerPath = arr[1]
-		mnt.writable = validMountMode(arr[2]) && arr[2] == "rw"
+		driver = p[0]
```

Comment:
Sorry, I don't get it: the driver name can be specified as part of the `-v` argument? I don't think it should.